### PR TITLE
refactor: move push click handling to processor

### PR DESCRIPTION
**Pull request by mrehan27**
helps: https://github.com/customerio/issues/issues/10830

### Changes

- Moved code out of `NotificationClickReceiverActivity` to `PushMessageProcessor` for reusability and better testing
- Added tests for `CustomerIOPushNotificationHandler`, `ModuleMessagingConfig` and `PushMessageProcessor`

### CODE REVIEW

**Review by Shahroz16 — COMMENTED** [review]
`getApplicationInfo(String, PackageManager.ApplicationInfoFlags)` lets maybe use this, since the other method is deprecated.

File: messagingpush/src/main/java/io/customer/messagingpush/activity/NotificationClickReceiverActivity.kt
Comment (levibostian) [inline messagingpush/src/main/java/io/customer/messagingpush/activity/NotificationClickReceiverActivity.kt:40]:
```suggestion
        if (data == null) {
            // This should never happen ideally
            logger.error("Intent is null, cannot process notification click")
            return 
        }      
```

I think the guard design pattern would be nice here instead of using `if/else`. Not a blocker, just nit-picky.

File: messagingpush/src/main/java/io/customer/messagingpush/processor/PushMessageProcessorImpl.kt
Comment (levibostian) [inline messagingpush/src/main/java/io/customer/messagingpush/processor/PushMessageProcessorImpl.kt:120]: Is there a chance that `trackMetrics()` would be called manually? If so, the `if (moduleConfig.autoTrackPushEvents)` should be moved into maybe `processNotificationIntent` function.

File: messagingpush/src/sharedTest/java/io/customer/messagingpush/CustomerIOPushNotificationHandlerTest.kt
Comment (levibostian) [inline messagingpush/src/sharedTest/java/io/customer/messagingpush/CustomerIOPushNotificationHandlerTest.kt:41]: Would it add value if there was another test function added where `given not valid payload, should not start Activity`?

**Review by levibostian — COMMENTED** [review]
Looking good! I am OK with approving, but with a big PR like this, I would prefer multiple reviews and approvals on this one. So, I'll let Shahroz help me and approve this one.

File: messagingpush/src/main/java/io/customer/messagingpush/activity/NotificationClickReceiverActivity.kt
Comment (mrehan27) [reply messagingpush/src/main/java/io/customer/messagingpush/activity/NotificationClickReceiverActivity.kt:40]: We unfortunately can't use guard here because it should finish activity in all cases before exiting from this method.

File: messagingpush/src/main/java/io/customer/messagingpush/activity/NotificationClickReceiverActivity.kt
Comment (mrehan27) [reply messagingpush/src/main/java/io/customer/messagingpush/activity/NotificationClickReceiverActivity.kt:45]: We can. But I avoided because I don't think this adds a lot of value. It mostly relies on Android OS lifecycle and couple of other methods that are already being tested. Also, it requires setting up Activity lifecycle for testing which I plan to improve later to gain more confidence in testing. However, if you think we should still cover this, I can write a test for it.

File: messagingpush/src/main/java/io/customer/messagingpush/processor/PushMessageProcessorImpl.kt
Comment (mrehan27) [reply messagingpush/src/main/java/io/customer/messagingpush/processor/PushMessageProcessorImpl.kt]: I don't want to pass this responsibility to `processNotificationIntent` actually, because I think `trackMetrics` should be responsible for checking this.

However, I rechecked and realized that latest versions of tests don't require these methods to be exposed, so I have updated the visibility. Also, looking at the code again, I feel like the method names should have been more descriptive after they were moved to `PushMessageProcessorImpl`, I have pushed changes after renaming them so they make more sense.

File: messagingpush/src/sharedTest/java/io/customer/messagingpush/CustomerIOPushNotificationHandlerTest.kt
Comment (mrehan27) [reply messagingpush/src/sharedTest/java/io/customer/messagingpush/CustomerIOPushNotificationHandlerTest.kt]: No actually. The activity should be started in all cases because this is being triggered when notification is clicked. So even if the payload is broken, Android OS still need something to start when user taps the notification. Whether we open anything next or not will be decided by `NotificationClickReceiverActivity`.

But I see the test name is confusing, I have updated it.

**Comment by mrehan27** [comment]
@Shahroz16 I can see many warnings on `getApplicationInfo`, but the part of code using it is not linked with this change and exist in current `main` branch. I agree we should fix this warning, but it should be done and tested in a separate PR.

**Comment by mrehan27** [comment]
@levibostian Most of the changes are actually copied from `NotificationClickReceiverActivity` to `PushMessageProcessorImpl` to facilitate testing. I avoid splitting into multiple PRs because except tests, all other changes are already been approved in #249. But I understand the concern, and will wait for both of you to approve before merging.

File: messagingpush/src/main/java/io/customer/messagingpush/activity/NotificationClickReceiverActivity.kt
Comment (Shahroz16) [reply messagingpush/src/main/java/io/customer/messagingpush/activity/NotificationClickReceiverActivity.kt:45]: I think it still alots value because the whole logic revolves around what flags to expect and what behaviour it would display.

**Review by Shahroz16 — APPROVED** [review]

**Review by levibostian — APPROVED** [review]

File: messagingpush/src/main/java/io/customer/messagingpush/activity/NotificationClickReceiverActivity.kt
Comment (mrehan27) [reply messagingpush/src/main/java/io/customer/messagingpush/activity/NotificationClickReceiverActivity.kt:45]: I have pushed more tests, hope it helps.

PS: Github is having issues in PRs, so it might take some time to show new changes.

**Pull request merged**
